// tlock-js/crypto/fp.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.fp12FromBigTwelve = exports.fp6FromBigSix = exports.fp2FromBigTuple = exports.Fp6 = exports.Fp2 = exports.Fp = void 0;
function fp2FromBigTuple(t) {
    return {
        c0: t[0],
        c1: t[1]
    };
}
exports.fp2FromBigTuple = fp2FromBigTuple;
function fp6FromBigSix(b) {
    return {
        c0: {
            c0: b[0],
            c1: b[1],
        },
        c1: {
            c0: b[2],
            c1: b[3],
        },
        c2: {
            c0: b[4],
            c1: b[5],
        }
    };
}
exports.fp6FromBigSix = fp6FromBigSix;
function fp12FromBigTwelve(b) {
    return {
        c0: {
            c0: {
                c0: b[0],
                c1: b[1],
            },
            c1: {
                c0: b[2],
                c1: b[3],
            },
            c2: {
                c0: b[4],
                c1: b[5],
            }
        },
        c1: {
            c0: {
                c0: b[6],
                c1: b[7],
            },
            c1: {
                c0: b[8],
                c1: b[9],
            },
            c2: {
                c0: b[10],
                c1: b[11],
            }
        }
    };
}
exports.fp12FromBigTwelve = fp12FromBigTwelve;
